File: src/bin/0073-set-matrix-zeroes.cpp
// @leetup=info id=73 lang=cpp slug=set-matrix-zeroes

#include <algorithm>
#include <iostream>
#include <map>
#include <set>
#include <tuple>
#include <vector>

// @leetup=code

class Solution {
 public:
  void setZeroes(std::vector<std::vector<int>>& matrix) {
    std::set<size_t> zero_r;
    std::set<size_t> zero_c;

    for (size_t r = 0; r < matrix.size(); ++r) {
      for (size_t c = 0; c < matrix[0].size(); ++c) {
        if (matrix[r][c] == 0) {
          zero_r.insert(r);
          zero_c.insert(c);
        }
      }
    }

    for (size_t r = 0; r < matrix.size(); ++r) {
      for (size_t c = 0; c < matrix[0].size(); ++c) {
        if (zero_r.find(r) != zero_r.end() || zero_c.find(c) != zero_c.end()) {
          matrix[r][c] = 0;
        }
      }
    }
  }
};
// @leetup=code

int main() {
  std::cout << "Hello, world!" << std::endl;
  return 0;
}

File: src/bin/0073-set-matrix-zeroes.py
# @leetup=info id=73 lang=python3 slug=set-matrix-zeroes



# @leetup=code


class Solution:
    def setZeroes(self, matrix: list[list[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        R = len(matrix)
        C = len(matrix[0])

        zero_r = set()
        zero_c = set()

        for r in range(R):
            for c in range(C):
                if matrix[r][c] == 0:
                    zero_r.add(r)
                    zero_c.add(c)

        for r in range(R):
            for c in range(C):
                if r in zero_r or c in zero_c:
                    matrix[r][c] = 0


# @leetup=code
